docs/proposals/set-ref.md

```diff
@@ -0,0 +1,144 @@
+
+Secondary goal of this effort is to improve performance of various controllers and schedulers, by removing the need for expensive lookup for all matching "controllers".
+
+# Non goals
+
+Cascading deletion is not a goal of this effort. Cascading deletion will use *ownerReferences*, which is a [separate effort](garbage-collection.md).
+
+There will be some interactions between set and owner references, they will have a similar orphaning/adoption behavior and *setReference* may be treated as one of *ownerReferencess*.
+
+# API and semantic changes
+
+There will be a new API field in the *ObjectMeta* in which we will store a *setReference*:
+
+```
+ObjectMeta {
+    …
+    setReference *SetReference
+    …
+}
+```
+
+```
+type SetReference struct {
+    // Version of the referent in group/version format.
+    APIVersion string
+    // Kind of the referent.
+    Kind string
+    // Name of the referent.
+    Name string
+    // UID of the referent.
+    UID types.UID
+    // should this SetReference be treated as a parent for cascading deletion
+    isParent bool
+}
+```
+
+Most controllers (all that manage collections of things defined by label selector) will have slightly changed semantics: currently controller owns an object if its selector matches object’s labels, but after introduction of *setReference* a controller will own an object iff selector matches labels and *setReference* points to it.
+
+If the owner's selector or owned object's labels change, the owning controller will be responsible for orphaning (clearing *setReference*) objects, after which adoption procedure (setting *setReference*) might occur, if another controller has a selector matching..
+
+User shouldn't set *setReference* by himself/herself, only system components should be responsible for updating this field. In the fullness of time an Admission Control plugin may block all updates to *setReference* signed by something different than Controller Manager.
+
+For debugging purposes we want to add an `adoptionTime` annotation which will keep the time of last ownership transfer.
+
+# Upgrade/downgrade procedure
+
+During system upgrade controllers will adopt objects that match their label selector using ordinary adoption procedure. Adoption, by default, will set *IsParent* to *false,* so the behavior of objects existing before the update won’t change (i.e. deleting owners won't cause cascading deletion - see OwnerReference design).
+
+In case of system downgrade controllers will ignore *setReference* and everything will work as it does in the old version of the system.
+
+In case of subsequent system upgrade GarbageCollector will clear invalid *setReferences* and normal adoption will take place afterwards.
+
+# Orphaning/adoption
+
+Controllers will orphan objects they own in two cases:
+* Change of label/selector causing selector to stop matching labels (executed by the controller)
+* Deletion of a controller with *orphaning=true* (executed by the garbage collector)
```

Comment:
I mean it looks odd to me if the reference is not even an ownerReference and the GC still needs to take care of it.